addons/crm/models/crm_team.py

```diff
@@ -582,6 +560,81 @@ def _allocate_leads_deduplicate(self, leads, duplicates_cache=None):
             'duplicates': leads_dup_ids,
         }
 
+    def _assign_and_convert_leads(self, force_quota=False):
+        """ Main processing method to assign leads to sales team members. It also
+        converts them into opportunities. This method should be called after
+        ``_allocate_leads`` as this method assigns leads already allocated to
+        the member's team. Its main purpose is therefore to distribute team
+        workload on its members based on their capacity.
+
+        This method follows the following heuristic
+            * Get quota per member
+            * Find all leads to be assigned per team
+            * Sort list of members per number of leads received in the last 24h
+            * Assign the lead using round robin
+                * Find the first member with a compatible domain
+                * Assign the lead
+                * Move the member at the end of the list if quota is not reached
+                * Remove it otherwise
+                * Move to the next lead
+
+        :param bool force_quota: see ``CrmTeam._action_assign_leads()``;
+
+        :return members_data: dict() with each member assignment result:
+          membership: {
+            'assigned': set of lead IDs directly assigned to the member;
+          }, ...
+
+        """
+        auto_commit = not getattr(threading.current_thread(), 'testing', False)
+        result_data = {}
+        commit_bundle_size = int(self.env['ir.config_parameter'].sudo().get_param('crm.assignment.commit.bundle', 100))
+        quota_per_member = {member: member._get_assignment_quota(force_quota=force_quota) for member in self.crm_team_member_ids}
+        counter = 0
+        leads_per_team = dict(self.env['crm.lead']._read_group(
+            [('user_id', '=', False), ('date_open', '=', False), ('team_id', 'in', self.ids)],
+            ['team_id'],
+            ['id:recordset'],
+        ))
+        for team, leads_to_assign in leads_per_team.items():
+            members_to_assign = list(team.crm_team_member_ids.filtered(lambda member:
```

Comment:
Note that we could begin the method with a first filter before fetching "leads_per_team", just to avoid doing a read_group on teams that have no members with stuff to assign. Like a "valid_teams" pre-computation, just to be sure we avoid fetching useless leads.